packet.go

```diff
@@ -66,29 +89,70 @@ func (p *Packet) IsRaw() bool {
 
 // Decode decodes the headers of a Packet.
 func (p *Packet) Decode() {
+
+	// Our go tests do not initialize p.Payload.
+	if p.Payload == nil {
+		p.Payload = p.Data
+	}
 	if p.IsRaw() == false {
-		p.Type = int(binary.BigEndian.Uint16(p.Data[12:14]))
-		p.DestMac = decodemac(p.Data[0:6])
-		p.SrcMac = decodemac(p.Data[6:12])
-		p.Payload = p.Data[14:]
-		switch p.Type {
-		case 0x8100:
-			//VLAN tag skip 802.1Q tag
-			p.Type = int(binary.BigEndian.Uint16(p.Data[16:18]))
-			p.Payload = p.Data[18:]
-		case 0x88A8:
-			//VLAN tag skip 802.1ad double tag
-			p.Type = int(binary.BigEndian.Uint16(p.Data[20:22]))
-			p.Payload = p.Data[22:]
-		}
+		// we want to set mac addresses of the Packet p using outermost
+		// ethernet frame. These are used to determine directionality of the
+		// captured packet.
+		p.decodeEther(true /* outerMac */)
 	} else {
 		p.Type = TYPE_IP
```

Comment:
It might be good to check for the ipv4 case first as that is the most frequent likely case. Also that will let you assign p.Type only once rather than with a default assignment first and revising it with the ipv6 case check. This is not important now but just an optimization.

Reply:
This is not the usual code path for us. We start with ethernet frame.